common_components/linux_compat/freertos/osal/queue.cpp

```diff
@@ -0,0 +1,79 @@
+/*
+ * SPDX-FileCopyrightText: 2021-2022 Espressif Systems (Shanghai) CO LTD
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <cstdint>
+#include <vector>
+#include <cstring>
+#include <queue>
+#include <mutex>
+#include <condition_variable>
+#include <atomic>
+#include "osal_api.h"
+
+template <class T>
+class Queue {
+public:
+    Queue(): q(), m(), c() {}
+    ~Queue() {}
+
+    void send(std::unique_ptr<T> t)
+    {
+        std::lock_guard<std::mutex> lock(m);
+        q.push(std::move(t));
+        c.notify_one();
+    }
+
+    std::unique_ptr<T> receive(uint32_t ms)
+    {
+        std::unique_lock<std::mutex> lock(m);
+        while (q.empty()) {
+            if (c.wait_for(lock, std::chrono::milliseconds(ms)) == std::cv_status::timeout) {
+                return nullptr;
+            }
+        }
+        std::unique_ptr<T> val = std::move(q.front());
+        q.pop();
+        return val;
+    }
+
+private:
+    std::queue<std::unique_ptr<T>> q;
```

Comment:
> Could you elaborate on why store a std::unique_ptr<T> and not only T?

Since I don't have to deal with `delete`-ing the item when retrieving it from the queue.

Reply:
One more reason for staying witht `std::unique_ptr<>` is to use move semantics.

Reply:
I'm using this move semantics a lot and it seems easier to template this using `std::unique_ptr<>` rather than creating our own move operators, IMO.